addons/sale_project/models/sale_order.py

```diff
@@ -88,56 +104,84 @@ def _action_confirm(self):
+        }
         return action
 
+    def action_create_project(self):
+        self.ensure_one()
+        if not self.order_line:
+            return {'type': 'ir.actions.act_window_close'}
+
+        sorted_line = self.order_line.sorted('sequence')
+        default_sale_line = next(sol for sol in sorted_line if sol.product_id.detailed_type == 'service')
+        return {
+            **self.env["ir.actions.actions"]._for_xml_id("project.open_create_project"),
+            'context': {
+                'default_sale_order_id': self.id,
+                'default_sale_line_id': default_sale_line.id,
+                'default_partner_id': self.partner_id.id,
+                'default_user_ids': [self.env.uid],
+                'default_allow_billable': 1,
+                'hide_allow_billable': True,
+                'default_company_id': self.company_id.id,
+                'generate_milestone': default_sale_line.product_id.service_policy == 'delivered_milestones',
+            },
+        }
+
     def action_view_project_ids(self):
         self.ensure_one()
-        view_form_id = self.env.ref('project.edit_project').id
-        view_kanban_id = self.env.ref('project.view_project_kanban').id
+        if not self.order_line:
+            return {'type': 'ir.actions.act_window_close'}
+
+        sorted_line = self.order_line.sorted('sequence')
+        default_sale_line = next(sol for sol in sorted_line if sol.product_id.detailed_type == 'service')
         action = {
             'type': 'ir.actions.act_window',
-            'domain': [('id', 'in', self.project_ids.ids)],
-            'view_mode': 'kanban,form',
             'name': _('Projects'),
+            'domain': ['|', ('sale_order_id', '=', self.id), ('id', 'in', self.project_ids.ids)],
             'res_model': 'project.project',
+            'views': [(False, 'kanban'), (False, 'tree'), (False, 'form')],
+            'view_mode': 'kanban,tree,form',
+            'context': {
+                **self._context,
+                'default_partner_id': self.partner_id.id,
+                'default_sale_line_id': default_sale_line.id,
+                'default_allow_billable': 1,
+            }
         }
         if len(self.project_ids) == 1:
-            action.update({'views': [(view_form_id, 'form')], 'res_id': self.project_ids.id})
-        else:
-            action['views'] = [(view_kanban_id, 'kanban'), (view_form_id, 'form')]
+            action.update({'views': [(False, 'form')], 'res_id': self.project_ids.id})
         return action
 
     def action_view_milestone(self):
         self.ensure_one()
         default_project = self.project_ids and self.project_ids[0]
-        default_sale_line = default_project.sale_line_id or self.order_line and self.order_line[0]
+        sorted_line = self.order_line.sorted('sequence')
+        default_sale_line = next(sol for sol in sorted_line if sol.is_service and sol.product_id.service_policy == 'delivered_milestones')
```

Comment:
We could maybe do a search instead of sorting and then find the first sol whom satisfies your condition. 🤔 

I guess it would be faster with a search. 🙂 

```suggestion
        default_sale_line = self.env['sale.order.line'].search([
            ('is_service', '=', True), 
            ('service_policy', '=', 'delivered_milestones')
        ], order='sequence', limit=1)
```

PS: I did not test this code. 🙂

Reply:
In fact, since we will not have many SOLs in a SO, my comment is not really better. You can ignore my comment. 😇